fix: Generate correct collection size when annotation_types.Len is used

priyankc opened the pull request:
## Description
Fixes a bug where annotation_types.Len can go over the described length

## Test
### Before the fix:
```
➜  polyfactory git:(workback-fix-710) pytest tests/test_collection_length.py
======================================= test session starts =======================================
platform darwin -- Python 3.12.10, pytest-8.4.0, pluggy-1.6.0
rootdir: /Users/priyank/OSS/polyfactory
configfile: pyproject.toml
plugins: hypothesis-6.135.4, Faker-37.3.0, cov-6.1.1, asyncio-1.0.0
asyncio: mode=Mode.AUTO, asyncio_default_fixture_loop_scope=function, asyncio_default_test_loop_scope=function
collected 80 items

tests/test_collection_length.py F.......................................................... [ 73%]
.....................                                                                       [100%]

============================================ FAILURES =============================================
___________________________________ test_annotated_type_length ____________________________________

    def test_annotated_type_length() -> None:
        @dataclass
        class Foo:
            foo: Annotated[List[int], at.Len(1)]
        class FooFactory(DataclassFactory[Foo]):
            __model__ = Foo
            __randomize_collection_length__ = True

        for _ in range(10):
            foo = FooFactory.build()
>           assert len(foo.foo) == 1, len(foo.foo)
E           AssertionError: 2
E           assert 2 == 1
E            +  where 2 = len([5068, 9979])
E            +    where [5068, 9979] = test_annotated_type_length.<locals>.Foo(foo=[5068, 9979]).foo

tests/test_collection_length.py:28: AssertionError
===================================== short test summary info =====================================
FAILED tests/test_collection_length.py::test_annotated_type_length - AssertionError: 2
================================== 1 failed, 79 passed in 0.53s ===================================
```

### After the fix
```
➜  polyfactory git:(workback-fix-710) ✗ pytest tests/test_collection_length.py
=============================================== test session starts ================================================
platform darwin -- Python 3.12.10, pytest-8.4.0, pluggy-1.6.0
rootdir: /Users/priyank/OSS/polyfactory
configfile: pyproject.toml
plugins: hypothesis-6.135.4, Faker-37.3.0, cov-6.1.1, asyncio-1.0.0
asyncio: mode=Mode.AUTO, asyncio_default_fixture_loop_scope=function, asyncio_default_test_loop_scope=function
collected 80 items

tests/test_collection_length.py ............................................................................ [ 95%]
....                                                                                                         [100%]

================================================ 80 passed in 0.48s ================================================
```
## Closes
Fixes #710

Rub1kCube commented on tests/test_collection_length.py:
```diff
@@ -9,9 +10,24 @@
 from polyfactory.factories import DataclassFactory
 from polyfactory.factories.pydantic_factory import ModelFactory
 
+from typing import Annotated
```
I'd suggest replacing the current import with `from typing_extensions import Annotated` for better compatibility.

Rub1kCube reviewed (commented):
Important workflow note: when creating a PR, make sure the target repository points to the main project, not the fork. Currently the commit is directed to your personal main branch.

priyankc commented on tests/test_collection_length.py:
```diff
@@ -1,6 +1,7 @@
 from enum import Enum
 from typing import Any, Dict, FrozenSet, List, Literal, Optional, Set, Tuple, get_args
 
+import annotated_types as at
```
done.

priyankc commented:
@Rub1kCube Thanks for the quick review. Addressed all the comments. I will raise the PR against the main project and will add you as reviewer. Will close this PR, so you can do another review on it.
